subprojects/model-core/src/integTest/groovy/org/gradle/api/provider/PropertyAssignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,405 @@
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class PropertyAssignmentIntegrationTest extends AbstractIntegrationSpec {
+
+    def "test Groovy object types assignment for #description"() {
+        given:
+        def buildFileDefinition = """
+            enum MyEnum {
+                YES, NO
+            }
+
+            class MyObject {
+                private String value
+                public MyObject(String value) {
+                    this.value = value
+                }
+                public String toString() {
+                    return value
+                }
+            }
+
+            abstract class MyTask extends DefaultTask {
+                @Internal
+                {inputDeclaration}
+
+                @TaskAction
+                void run() {
+                    if (input instanceof DirectoryProperty) {
+                        println("Result: " + input.get().asFile.name)
+                    } else if (input instanceof File) {
+                       println("Result: " + input.name)
+                    } else if (input instanceof Property) {
+                        println("Result: " + input.get())
+                    } else {
+                        println("Result: " + input)
+                    }
+                }
+            }
+
+            tasks.register("myTask", MyTask) {
+                input = {inputValue}
+            }
+        """
+
+        when:
+        buildFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "$eagerInputType input")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        runAndAssert("myTask", eagerResult)
+
+        when:
+        buildFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "abstract $lazyInputType getInput()")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        runAndAssert("myTask", lazyResult)
+
+        where:
+        description                                     | eagerInputType | lazyInputType       | inputValue                               | eagerResult                          | lazyResult
+        "T = T"                                         | "String"       | "Property<String>"  | '"hello"'                                | "Result: hello"                      | "Result: hello"
```

Comment:
🤔 How about removing `Result: ` from `eagerResult` and `lazyResult`, and add the prefix in the test code instead? That would shorten the lines here, too.

Reply:
Done